src/main/java/seedu/address/model/company/UniqueCompanyList.java

```diff
@@ -0,0 +1,159 @@
+package seedu.address.model.company;
+
+import static java.util.Objects.requireNonNull;
+import static seedu.address.commons.util.CollectionUtil.requireAllNonNull;
+
+import java.util.Iterator;
+import java.util.List;
+
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import seedu.address.model.company.exceptions.CompanyNotFoundException;
+import seedu.address.model.company.exceptions.DuplicateCompanyException;
+
+/**
+ * A list of companies that enforces uniqueness between its elements and does not allow nulls.
+ * A company is considered unique as deemed by the <Code>Company.equals(Company)</Code>.
+ * Additions and deletions use this logic to determine what can be added and removed.
+ */
+public class UniqueCompanyList implements Iterable<Company> {
+    private final ObservableList<Company> internalList = FXCollections.observableArrayList();
+    private final ObservableList<Company> internalUnmodifiableList =
+            FXCollections.unmodifiableObservableList(internalList);
+
+    /**
+     * Returns true if the list contains a company.
+     *
+     * @param toCheck Company to be checked.
+     * @return true if list already contains company.
+     */
+    public boolean contains(Company toCheck) {
+        requireNonNull(toCheck);
+        return internalList.stream().anyMatch(toCheck::equals);
+    }
+
+    /**
+     * Adds a company to the list.
+     * Company must not exist in the list.
+     *
+     * @param toAdd Company to be added.
+     */
+    public void add(Company toAdd) {
+        requireNonNull(toAdd);
+        if (contains(toAdd)) {
+            throw new DuplicateCompanyException();
+        }
+        internalList.add(toAdd);
+    }
+
+    /**
+     * Replaces the specified company with an edited one.
+     *
+     * @param target Company to be replaced in list.
+     * @param editedCompany Company to replace the other.
+     */
+    public void setCompany(Company target, Company editedCompany) {
+        requireAllNonNull(target, editedCompany);
+
+        int index = internalList.indexOf(target);
+        if (index == -1) {
+            throw new CompanyNotFoundException();
+        }
+
+        if (!target.equals(editedCompany) && contains(editedCompany)) {
```

Comment:
Potentially need to change `Company::equals` to `Company::isSameCompany` so that the `DuplicateCompanyException` is correctly used.

Reply:
changed